### ISSUE TITLE
Bad network connection sometimes gives double entries

### ISSUE THREAD

**Issue by txcom2003** [odoo/odoo#54132]
Impacted versions: 12.0

Steps to reproduce:

1. Prepare 2 kind of connection (using Network Link Conditioner, or other) :

- Good Network Connection (Wifi)
- Bad Network Connection (low packet drop rate for uplink and high packet drop rate for downlink)

2. Create new entry (ex: create 1 new vendor bill line entry from existing vendor bill) while using good network connection, do not press save button yet
3. Client switch the connection to use bad network connection
4. Click save button
5. Wait until connection timeout
6. Switch to Good network connection

Current behavior:
Sometimes generate multiple entries

Expected behavior:
generate only 1 entry

Video/Screenshot link (optional):
[https://www.dropbox.com/s/bdj2qhgp9m3l019/odoo-chrome-double-request-02.mov?dl=0](url)

**Comment by txcom2003** [comment]
It seems that the browser retry the request when certain condition happened as mentioned in the following posts:
- https://stackoverflow.com/questions/15155014/inconsistent-browser-retry-behaviour-for-timed-out-post-requests
- http://geek.starbean.net/?p=393

this problem can cause duplicate entries and mess up our inventory and accounting data

I haven't used runbot infrastructure before, how to use it ? currently I only see green door icon only

**Comment by pedrobaeza** [comment]
Yes, the green one (Odoo changed the color recently). That link seems to refer to node.js.

**Comment by txcom2003** [comment]
Okay, I will try to reproduce the problem in runbot infrastructure

It’s the browser behaviour that resend request automatically when certain condition happened, so it can affect any server, not only node.js. what we can do is to make sure that the same request is not processed twice at the server side. one example to address this problem is to generate unique request id and send it together with the user input data to server, when the request arrived at server side, server will check whether the request id is already processed or not. if it already processed then server should not process it again.

**Comment by pedrobaeza** [comment]
Which browser? I use Firefox and Chrome without noticing that behavior.

**Comment by txcom2003** [comment]
i used chrome to reproduce the issue in the following videos
https://www.dropbox.com/s/bdj2qhgp9m3l019/odoo-chrome-double-request-02.mov?dl=0

and it also happened in firefox

**Comment by pedrobaeza** [comment]
Well, not sure if that can qualify for a bug in such extreme condition that you needed to reproduce through artificial conditions. @odony who can qualify this?

**Comment by odony** [comment]
@txcom2003 tricky case, difficult to reproduce and analyze...  Here are some thoughts.

From what I understand, some browsers tend to implement a retry mechanism that is closer to what is specified in the old RFC 2616, [section 2.8.4](https://tools.ietf.org/html/rfc2616#section-8.2.4) (basically: anything can be retried if the connection is lost without response), rather than the newer spec from RFC 2730, [section 6.3.1](https://tools.ietf.org/html/rfc7230#section-6.3.1) (which only allows retrying *idempotent* requests, so normally not a POST). That's annoying, but we can't do much.

That old RFC states that the retry can *only* occur if the connection is abruptly dropped without response. An origin server (Odoo or its reverse proxy) would never drop the connection without response, it would give a 400 or 500 error code even after a timeout, depending on the stage of processing. So if packets are lost in transmission the connection would theoretically just stay open forever, unless some device in the middle decides to close it.

Is there any network layer between you and the server that implements some safety timeout? In your wireshark session, if you look at the TCP level you might be able to notice the RST packet closing the connection, and see which device sent it. That could help diagnose the issue, and could perhaps be fixed by changing the timeouts on that device/layer, making an exception in the timeout rules, or making the equipment throw a proper HTTP error instead of dropping the connection.

Normally when your connection changes the browser will notice and abort all requests. You will see them fail with ERR_NETWORK_CHANGED (Chrome). It doesn't happen for you - for some reason the network stack believes you're still on the same network link. Perhaps you could try to have different IPs on the "bad" and the "good" network to help the browser detect the change.

On the Odoo side, implementing a shared state mechanism to remember the list of already processed requests, just for this extreme and rare use case feels overkill. I don't think there's any chance we'll do it in the official version. It's non-trivial to cover all cases include HTTP forms in addition to JSON-RPC calls, and it could have some undesired side-effects (sometimes replaying a call is idempotent, and even desirable instead of throwing an error)

Note: the internal response timeout of the browser should not cause this behavior, but it's interesting to note that Firefox has a setting for tuning it via about:config (`network.http.response.timeout`). By making this timeout shorter than the time it takes to trigger the "hidden retry" you might also be able to eliminate the problem.

**Comment by pedrobaeza** [comment]
Thanks Olivier as always for the detailed explanation. I think this can be closed then, isn't it?

**Comment by txcom2003** [comment]
Hi Pedro, Olivier

Here I have reproduced double request at runbot infrastructure, can you take a look ? 
https://www.dropbox.com/s/2b3jxwdudkk3vb8/double-request-runbot-01.mov?dl=0
https://www.dropbox.com/s/pebeiq5ssjkt1z6/double-request-runbot-02.mov?dl=0

I'm still analyzing it, maybe there is some insight from you guys.
yeah it's difficult to reproduce and analyze, but it still can be happened (and actually happened to our team members several times, so we decide to investigate it since it can also can be happened to our users).
> 
> Is there any network layer between you and the server that implements some safety timeout? In your wireshark session, if you look at the TCP level you might be able to notice the RST packet closing the connection, and see which device sent it. That could help diagnose the issue, and could perhaps be fixed by changing the timeouts on that device/layer, making an exception in the timeout rules, or making the equipment throw a proper HTTP error instead of dropping the connection.
> 
> Note: the internal response timeout of the browser should not cause this behavior, but it's interesting to note that Firefox has a setting for tuning it via about:config (`network.http.response.timeout`). By making this timeout shorter than the time it takes to trigger the "hidden retry" you might also be able to eliminate the problem.

from the other day video, It seems that the opposite of client (means: server) that sent the reset signal (the source ip indicate that) before the client retry the request (i though it should be from client side)

So you mean we need to shorten the timeout ? we set our server timeout quite long (5~10 minutes) since our users need to generate reports that can take several minutes to complete. How we can handle this ? any idea ?

**Comment by vishalp-cd** [comment]
Hello @odony,

I concur with your findings. I am experiencing a similar issue, namely, the triggering of two SMS messages simultaneously in our production environment.
